### PR TITLE
Add pred m to z not m

### PR DESCRIPTION
**Pull request by Rick-Methot-NOAA**
## What issue(s) does this PR address? Describe and add issue numbers, if applicable.

Link issue(s) here: #88 and #235 

## What tests have been done? Upload any model input files created for testing in a zip file, if possible.
extensive look at all test models, special testing using the predator model:  vendance.
## Check which is true. This PR requires:

- [ ] no further changes to r4ss
- [ ] no further changes to the manual
- [ ] no further changes to SSI (the SS3 GUI)
- [ ] no further changes to the stock synthesis change log (new features, bug reports)

## Describe any changes in r4ss/SS3 manual/SSI/change log that are needed (if not checked):
natural mortality report table has been reformatted
technical description of these changes is in attached doc
[natM tech doc.txt](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/8635992/natM.tech.doc.txt)

### CODE REVIEW

File: SS_write_report.tpl
Comment (k-doering-NOAA) [inline SS_write_report.tpl:3304]: remove "is", I think, to make this statement grammatically correct. Also, Are we using M1 and M2 throughout? I thought we were using something like M and M_pred ?? It would be good to be consistent.

File: SS_param.tpl
Comment (k-doering-NOAA) [inline SS_param.tpl:273]: question, style: should there be spaces between the indices? I think so...

many lines in this section are also missing `;` at the end of the line

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:156]: checking that || is right and shouldn't be &&. 
Also, style: missing spaces. `do_once > 0`

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:282]: style: missing spaces. `natM(t, 0);`

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:289]: style: missing curly braces for the if statment.

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:901]: style: missing curly braces around if statement. next if statement missing braces as well.

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:910]: style: missing spaces `for (p = 1; p <= pop; p++)`

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:912]: style: missing spaces. `int s1 = (p - 1) * nseas + s;`

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:1206]: what is Z_rate vs Zrate2? It would be nice to give these more meaningful names.

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:1238]: There should be an error that prevents users trying to use Pope and pred_M2 together, not sure if there is already or not.

File: SS_popdyn.tpl
Comment (k-doering-NOAA) [inline SS_popdyn.tpl:1343]: missing curly brace right after for statement.

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1339]: Is this left over from debugging? Or add a more informative echo input msg.

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1344]: style: add spaces `natM(t, 0)...`

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1347]: style: add space. `...natM(t, 0);`

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1351]: style: add spaces. `if (N_pred > 0)`

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1371]: style: add space. `natM(t, p, gpi)...`

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1385]: left over from debugging?

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1381]: style: add spaces.

Also, not sure if this is left over from debugging statement or not?

File: SS_prelim.tpl
Comment (k-doering-NOAA) [inline SS_prelim.tpl:1377]: If this is meant to help users, maybe a name more informative than surv would be helpful...I can't tell from the msg what is going on here.

File: SS_readdata_330.tpl
Comment (k-doering-NOAA) [inline SS_readdata_330.tpl:212]: style: add space `(1, Nfleet)` I think? Missing `;` at end of line?

Also needed on the two following lines.

File: SS_readdata_330.tpl
Comment (k-doering-NOAA) [inline SS_readdata_330.tpl:265]: style: missing space?

File: SS_readdata_330.tpl
Comment (k-doering-NOAA) [inline SS_readdata_330.tpl:282]: style: missing spaces. `predator_area(0, N_pred) = f;` 
Also needed on following line.

File: SS_readstarter.tpl
Comment (k-doering-NOAA) [inline SS_readstarter.tpl:52]: special_flag doesn't strike me as an informative name...maybe change to something more informative?

File: SS_selex.tpl
Comment (k-doering-NOAA) [inline SS_selex.tpl:4]: It looks like for some reason a bunch of additional lines were added into SS_selex.tpl. Not a huge deal, but maybe in the future don't commit files where changes weren't made intentionaly (I know it is easy to forget this with a big diff, though!)

File: SS_write_report.tpl
Comment (k-doering-NOAA) [inline SS_write_report.tpl:3302]: style: add spaces. `if (N_pred > 0)`

**Comment by k-doering-NOAA** [comment]
@Rick-Methot-NOAA , sorry, I noted after I did this review that @kellijohnson-NOAA was included as the assignee! Maybe in the future we should just request reviews instead of adding an assignee? I'm not sure assignee makes sense with our workflow, but it doesn't indicate to me that a review is wanted from this person...

**Comment by Rick-Methot-NOAA** [comment]
Thanks Kathryn for that thorough review.  It will take me a while to go through all of them, so I'll start with overall response:
1.  white space style.  It is very difficult to remember to put space around everything.  This is something that we should rely on the styler to add later.
2.   missing {} - agree that code is more explicit if they are always used, but in many instances the compiler does not need them. Much of what you noted was already there.
3. variable names.  I tried to lay out the logic  in the text document.  natM_M1(s  is calculated in biofxn and produces an ephemeral version of age-specific M that is immediately transferred into natM(t.  Then the overall level of predation for each predator is pred_M2.  So pred_M2 for predators is equivalent to Hrate for fishing fleets.   Neither pred_M2 or Hrate are age-specific and only inherit age-specificity when used in conjunction with the selectivity of that fleet or predator.  Age-specific predation gets added onto natM to get the total M (e.g. M1+m2).  Then Zrate is age-specific and based on natM + sum(Hrate*selec)
4. reviewers - I added Kelli to bring this to her attention, but I agree with you for our team style specific reviewers are less necessary.  However, having a reviewer named does make it more clear that someone needs to review before merging.
5. "special_flag" is for my special purposes while developing.  Leave it in else I need to create it anew.  I use it to do debugging write statements under particular situations.

**Comment by kellijohnson-NOAA** [comment]
I have this on my task list for today, though I don't know if I will be able to finish it b/c it might take me a while to get up to speed with all of the changes.

**Comment by Rick-Methot-NOAA** [comment]
here is an example that includes a predator in a 4 season empirical wt-at-age model.
[predator.zip](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/8642280/predator.zip)

**Comment by k-doering-NOAA** [comment]
> Thanks Kathryn for that thorough review. It will take me a while to go through all of them, so I'll start with overall response:
> 
> 1. white space style.  It is very difficult to remember to put space around everything.  This is something that we should rely on the styler to add later.
> 2. missing {} - agree that code is more explicit if they are always used, but in many instances the compiler does not need them. Much of what you noted was already there.
> 3. variable names.  I tried to lay out the logic  in the text document.  natM_M1(s  is calculated in biofxn and produces an ephemeral version of age-specific M that is immediately transferred into natM(t.  Then the overall level of predation for each predator is pred_M2.  So pred_M2 for predators is equivalent to Hrate for fishing fleets.   Neither pred_M2 or Hrate are age-specific and only inherit age-specificity when used in conjunction with the selectivity of that fleet or predator.  Age-specific predation gets added onto natM to get the total M (e.g. M1+m2).  Then Zrate is age-specific and based on natM + sum(Hrate*selec)
> 4. reviewers - I added Kelli to bring this to her attention, but I agree with you for our team style specific reviewers are less necessary.  However, having a reviewer named does make it more clear that someone needs to review before merging.
> 5. "special_flag" is for my special purposes while developing.  Leave it in else I need to create it anew.  I use it to do debugging write statements under particular situations.

Hi, Rick, thanks for all these responses! Certainly, feel free to not address comments I made that you don't agree with, as I don't want perfect to be the enemy of good. A few things, though:
1. I agree with using the styling tool to fix whitespace, but given it still isn't totally automated, I think it would be good to avoid introducing poor styling into main when possible. I can help make these changes, but I thought it might be an opportunity to help you remember the whitespace rules better? If you don't see it that way, though, I understand!
2. Adding in the {} makes the code more readable, even though the compiler doesn't need them. I can make these changes, if you would like. I think as we work on code, if we run into bad style we should fix it on the lines being modified?
3. Thanks for the tech document!
4. I see - Perhaps discussing how we use the fields on PRs would be something worth checking on.
5. Ah, got it! Maybe a comment where it is defined could make this clear?

File: SS_write_report.tpl
Comment (Rick-Methot-NOAA) [reply SS_write_report.tpl:3308]: let's check style guide on this.  There are many places in the code where 2 or more nested for() loops share the same {}

File: SS_readdata_330.tpl
Comment (Rick-Methot-NOAA) [reply SS_readdata_330.tpl:265]: readdata and readcontrol were not automatically styled

File: SS_benchfore.tpl
Comment (Rick-Methot-NOAA) [reply SS_benchfore.tpl:496]: I think some extra line endings got introduced by the styler.

Comment (k-doering-NOAA) [reply SS_benchfore.tpl:496]: oh, weird! @nschindler-noaa did you know about this?

**Comment by Rick-Methot-NOAA** [comment]
Most recommended changes were implemented.
array natM_M1 has been removed.  Now, M1 is stored directly into the 0'th row of natM().  Updated tech doc attached here.
[natM tech doc.txt](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/8653633/natM.tech.doc.txt)

**Comment by Rick-Methot-NOAA** [comment]
@k-doering-NOAA  - I hope I addressed your review requests adequately.  Did you get the vendance example for testing?

**Comment by k-doering-NOAA** [comment]
Thanks @Rick-Methot-NOAA , I will take look by Wed at the very latest. I see the example files, thank you!

**Comment by k-doering-NOAA** [comment]
@Rick-Methot-NOAA , did the changes get pushed up to github? I'm not seeing new commits...

File: SS_param.tpl
Comment (k-doering-NOAA) [reply SS_param.tpl:273]: Maybe we can discuss this at our next ss3 meeting, since we don't have a set style guide.

File: SS_write_report.tpl
Comment (k-doering-NOAA) [reply SS_write_report.tpl]: opened this in an issue #311

**Comment by k-doering-NOAA** [comment]
Good call @Rick-Methot-NOAA! 

I've been prioritizing the admb 13 testing over this, but I can course correct and work on reviewing this issue sooner, if needed. The code looks good, but I still wanted to do some model runs.

**Comment by Rick-Methot-NOAA** [comment]
K - stick with the console project.  I can create a suitable test for area-predators.

**Comment by Rick-Methot-NOAA** [comment]
I modified the vendance example to have areas, then assigned the predator to either area 1 or area 2.  The results confirm that predator-by-area is working as expected.  I believe this PR is ready to be merged into main.
[predator_area.zip](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/8789357/predator_area.zip)
Note that one of the recent commits has created a conflict in ss_selex.tpl

**Comment by k-doering-NOAA** [comment]
Thanks @Rick-Methot-NOAA! Luckily this merge conflict was pretty easy to resolve.

@kellijohnson-NOAA, do you still think we should set up a test to compare using predM with using the old way as just a fishing fleet and making sure the models should still get the same Z? Or would you support merging it into main given the spatial example Rick has created?

**Comment by Rick-Methot-NOAA** [comment]
The test of predator as fishing fleet was done when predator fleet was first created last fall with 3.30.18.  This new commit makes it area-specific, but in doing so it needed to change internal indexing of natM.  Since that internal indexing change has not perturbed anything else in the model, I hope we are good to go.

**Review by k-doering-NOAA — APPROVED** [review]
Sounds good @Rick-Methot-NOAA ! I approved the changes, so if you are ready merge into main, go ahead.

**Pull request merged**
